enrollment_predictions/models/auto_regressor_dec_tree.py

```diff
@@ -3,52 +3,123 @@
+    for _, row in data.iterrows():
+        year = row["year"]
+        terms = row["terms"]
+        
+        for term_data in terms:
+            term = term_data["term"]
+            
+            course_enrollments = {}
+            for course_data in term_data["courses"]:
+                course = course_data["course"]
+                sections = course_data["sections"]
+                
+                
+                total_enrollment = 0
+                for section_data in sections:
+                    if section_data["num"].startswith('A'):
+                        total_enrollment += section_data["enrolled"]
+                    num_seats = section_data["num_seats"]
+                    professor = section_data["professor"]
+                
+                course_enrollments[course] = total_enrollment
+                
+            for course, enrollment in course_enrollments.items():
+                rows.append({
+                    "year": year,
+                    "term": term,
+                    "course": course,
+                    "num_seats": num_seats,
+                    "professor": professor,
+                    "enrolled": enrollment
+                })
+    
+    return pd.DataFrame(rows)
+
+
+def data_preprocessing(data):
+    data = flatten_data(data)
+    
+    # Filter data
+    data = data[data['course'].str.startswith(('SENG', 'CSC', 'ECE'))]
+    
+    # Create unique identifier for each course offering
+    data['offering'] = data['course'] + "-" + data['year'].astype(str) + "-" + data['term'].astype(str)
+
+    # Aggregate total enrollment for each offering
+    data = data.groupby('offering').agg({
+        'year': 'first',
+        'term': 'first',
+        'course': 'first',
+        'num_seats': 'sum',
+        'enrolled': 'sum',
+        'professor': lambda x: ', '.join(x.unique())
+    }).reset_index()
+
     return data
 
+
 def feature_engineering(data, window_sizes):
-    course_groups = data.groupby('Course')
+    data = data.copy()
+    course_groups = data.groupby('offering')
 
     for win in window_sizes:
-        data['mean_prev_{}'.format(win)] = course_groups['Enrolled'].transform(lambda x: x.shift().rolling(window=win).mean())
-        data['median_prev_{}'.format(win)] = course_groups['Enrolled'].transform(lambda x: x.shift().rolling(window=win).median())
-        data['min_prev_{}'.format(win)] = course_groups['Enrolled'].transform(lambda x: x.shift().rolling(window=win).min())
-        data['max_prev_{}'.format(win)] = course_groups['Enrolled'].transform(lambda x: x.shift().rolling(window=win).max())
-        data['std_prev_{}'.format(win)] = course_groups['Enrolled'].transform(lambda x: x.shift().rolling(window=win).std())
-        data['ewm_{}'.format(win)] = course_groups['Enrolled'].transform(lambda x: x.shift().ewm(span=win).mean())
+        data.loc[:, 'mean_prev_{}'.format(win)] = course_groups['enrolled'].transform(lambda x: x.shift().rolling(window=win).mean())
```

Comment:
For some reason feature engineering does not compute the correct values. All new features appear to be NaN. 
Please write the output to a file and check result manually in case it's me doing something wrong when calling the function.